src/main/java/frc/robot/subsystems/PowerMonitor.java

```diff
@@ -0,0 +1,155 @@
+    }
+
+    /**
+     * @param channel to unmonitor
+     * @return boolean representing if removal from monitoring set was successful
+     */
+    public boolean unlogChannel(int channel) {
+        return monitoringChannels.remove(channel);
+    }
+
+    /**
+     * Runs action when channel current draw reaches a threshold. <br>
+     * Note, when channel reaches current draw you will have to re-register the event
+     *
+     * @param channel          to monitor
+     * @param currentThreshold current threshold before runnable is called
+     * @param runnable         action to preform when current is reached
+     * @return if addition to currentEvent list was successful
+     */
+    public boolean onCurrentAboveThreshold(int channel, double currentThreshold, Runnable runnable) {
+        if (channel < pdh.getNumChannels()) {
+            return maxCurrentEvents.add(new ChannelCurrentEvent(channel, currentThreshold, runnable));
+        }
+        return false;
+    }
+
+    /**
+     * Runs action when channel current draw goes below a threshold. <br>
+     * Note, when channel goes under min current draw you will have to re-register the event
+     *
+     * @param channel          to monitor
+     * @param currentThreshold current threshold before runnable is called
+     * @param runnable         action to preform when current is reached
+     * @return if addition to currentEvent list was successful
+     */
+    public boolean onCurrentBelowThreshold(int channel, double currentThreshold, Runnable runnable) {
+        if (channel < pdh.getNumChannels()) {
+            return minCurrentEvents.add(new ChannelCurrentEvent(channel, currentThreshold, runnable));
+        }
+        return false;
+    }
+
+    @Override
+    public void periodic() {
+        callCurrentMaxEvents();
+        callCurrentMinEvents();
+        monitoringChannels.forEach(c -> {
+            Logger.logDouble(loggingBase + "channel" + c + "Current", getChannelCurrent(c), Constants.ENABLE_LOGGING);
+        });
+        Logger.logDouble(loggingBase + "totalCurrent", getTotalCurrent(), Constants.ENABLE_LOGGING);
+        Logger.logDouble(loggingBase + "totalPower", getTotalPower(), Constants.ENABLE_LOGGING);
+        Logger.logDouble(loggingBase + "totalEnergy", getTotalEnergy(), Constants.ENABLE_LOGGING);
+        Logger.logDouble(loggingBase + "voltage", getVoltage(), Constants.ENABLE_LOGGING);
+        Logger.logBoolean(loggingBase + "haveBrownedOut", getFaults().Brownout, Constants.ENABLE_LOGGING);
+    }
+
+    private double getVoltage() {
+        return pdh.getVoltage();
+    }
+
+    private void callCurrentMaxEvents() {
+        List<ChannelCurrentEvent> events = maxCurrentEvents.stream().filter(event -> {
+            int channel = event.getChannel();
+            return getChannelCurrent(channel) > event.getCurrentThreshold();
+        }).toList();
+        events.forEach(e -> e.getAction().run());
+        maxCurrentEvents.removeAll(events);
```

Comment:
Instead of a `Runnable` you can get a BiFunction<channel, current, Boolean> that will return `true` if it cares to stay registered. This will save the need to re-register every time.

Reply:
done